platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -1123,36 +1158,32 @@ public void restart(ComponentState selected) {
         // If this configuration belongs to the select version, queue ourselves up for traversal.
         // If not, then remove our incoming edges, which triggers them to be moved across to the selected configuration
         if (component == selected) {
-            if (!evicted) {
+            if (replacement == null) {
                 resolveState.onMoreSelected(this);
                 return;
             }
         }
-        if (!incomingEdges.isEmpty()) {
-            restartIncomingEdges();
-        }
+        restartIncomingEdges();
     }
 
+    /**
+     * Called on losing nodes after conflict resolution to retarget their existing incoming
+     * edges to the winning node. This method must be called after any relevant state is updated
+     * so that retargeting chooses the correct new target node.
+     */
     private void restartIncomingEdges() {
         if (incomingEdges.size() == 1) {
             EdgeState singleEdge = incomingEdges.get(0);
             singleEdge.retarget();
-        } else {
+        } else if (incomingEdges.size() > 1){
             for (EdgeState edge : new ArrayList<>(incomingEdges)) {
                 edge.retarget();
             }
         }
-        // TODO: Restarting incoming edges should ensure they are pointing to the correct node.
-        // If they end up pointing to us after restart, we should not remove them.
-        clearIncomingEdges();
-    }
 
-    private void clearIncomingEdges() {
```

Comment:
Renamed to `removeAllIncomingEdges`, and move up next to `addIncomingEdge` and `removeIncomingEdge`